### ISSUE TITLE
What are the next steps for Unreal Tournament 2004?

### ISSUE THREAD

**Issue by tuxayo** [dreamer/luxtorpeda#45]
* Is this project listed as [a candidate](https://github.com/dreamer/luxtorpeda/wiki/Game-engines) already?: *YES*
* Project link: 
* Project license: non-libre game
* Does it work on 64-bit only Linux: *MAYBE* (without multi-lib right?, probably not)
* Is source code available?: *NO* non-libre game
* Is source code stored in Git?: *NO* non-libre game

## Additional information
- if needed I can gift a steam copy of the game to a luxtorpeda developper
- i think I still have the retail DVD containing the Linux binaries

**Comment by dreamer** [comment]
AppId: https://steamdb.info/app/13230/ for future reference.

Steam version of UT2004 is waiting on my wishlist for next sale. Regarding next steps: I haven't investigated into it in depth yet, but overall next steps would be:

- Locating and backing up all available Linux patches and downloadable binaries. I hope files in here will be enough: https://www.liflg.org/?catid=6&gameid=17 but likely not (I tried UT99 and there were newer patches needed to make local Linux version run). Looking up installers prepared by someone on Lutris is helpful here. Also, no cracks, obviously.
- Making sure the game actually runs on Linux still. According to PCGW the game was released for amd64 (and Loki installer does provide amd64 files), therefore I think focus should be on 64-bit version. The roadblocks in here are likely: old ABI for libstdc++ and glibc. This link might provide some helpful info: https://help.ubuntu.com/community/Games/Native/UnrealTournament2004
- Making the game run not only on local machine, but also in Steam Runtime environment. Goal is not to distribute some hacky version - the goal is to make easily maintainable package, that will work for majority of Linux users. I think it might involve ripping some additional libraries provided by Steam Runtime's mirror of Ubuntu 12.04, that are not included in Steam Runtime itself nowadays and include them with the Luxtorpeda package. Also, the game needs to be tested with SDL1-2 compat library ([link](https://www.patreon.com/posts/project-sdl12-25321792)) and testing results/bugs reported to SDL2 project. Initial testing by SDL devs indicate, that the game works ok with the compat lib on X11 and Wayland.
- Making sure which files from Steam version need to be patched (UT* patches tend to span multiple files and Steam version is probably already pre-patched to the latest available Windows version) - we don't want to overwrite files and render the game unplayable in Proton for example.
- Creating Luxtorpeda package repo on GitLab and bringing all this stuff together to provide simple tarball package in Luxtorpeda format. We can store binary files on GitLab using git-lfs.

If it looks like non-trivial amount of work… well, that's because it is ;)

At the moment I am preoccupied with making [DOSBox, *erhm*, work](https://www.vogons.org/viewtopic.php?f=32&t=69497) and backlog of tasks for Luxtorpeda projects is really, really long already, so I wouldn't mind if someone else started the packaging process for UT2004 instead.

**Comment by tuxayo** [comment]
Thanks a lot for such an analysis! That's indeed quite some work.

> Steam version of UT2004 is waiting on my wishlist for next sale.

I can take care of that ;) Only UT2004 or other games from this pack? https://isthereanydeal.com/game/unrealdealpack/info/#/
Like UT99 and Unreal Gold

**Comment by rcgordon** [comment]
Hi, @tuxayo pointed me to this bug report.

Here is the latest version of UT2004 for Linux, 3369.2:

http://treefort.icculus.org/ut2004/ut2004-lnxpatch3369-2.tar.bz2

I _think_ you can just unpack that into a Windows install and go: it has both 32 and 64-bit binaries. You might need to mess with System/Default.ini files if not included in the tarball, so you don't fail trying to use the Direct3D renderer. At some point with Unreal Engine 2, we had the Linux/Mac builds smart enough to say "ah, this was clearly a Windows .ini, replace WinDrv and D3DDrv with SDLDrv and OpenGLDrv," but I don't remember if UT2004 did that, and you'll probably want to mess with the SDL and GL settings beyond choosing those drivers anyhow.

I assume the Steam version ships with the Epic Megapack, but if not, use this instead to get the bonus content (it also contains the 3369.2 binaries)...this was a free download from Epic, so I assume it's safe to redistribute the extra content even if Steam didn't.

http://treefort.icculus.org/ut2004/ut2004megapack-linux.tar.bz2

I haven't tried these binaries in a long time, and I don't know what limitations they'll have. They come from December 2005, and we probably had libstdc++ nastiness with them.

It would be worth installing the game from a retail disc (there's a Linux installer on the retail Windows DVD and CDs!), patching it to the latest, and diffing it against whatever's on Steam, just so we know what we have and don't.

--ryan.

**Comment by MicrosoftTakeover** [comment]
This may be of some use: https://github.com/alanjjenkins/archlinux-package-ut2004-steam

**Comment by d10sfan** [comment]
Thanks for the links and guidance on that! It looks like the Steam version includes the megapack maps, so that makes things a bit easier. I have a package (https://gitlab.com/d10sfan/luxtorpeda-ut2004) that looks to work well from a clean steam install. There is an issue with seemingly luxtorpeda finding the right working directory, which I've expanded on below.

It'll ask you with Zenity for your cd key and then writes that and does symlinks to a new directory, so the original Windows data files don't get messed with. Then, it'll do a few symlinks for some libraries, and launch the game. The first time load can take a while when it gets things ready, but after that it loads very quickly.

It uses sdlcl for sdl 1.2 support, which is included. For now, it will check if libstdc++.so.5 is installed on the user's system, and if not, it will prompt with a Zenity error. It will also check for openal, which it'll link to the proper place. If both are installed and the cd key is entered properly, it should then run without any extra user input, I figure in the future we could package that in some way along with the rest.

@dreamer Something I noticed that seems specific to this one:

When I launch a game through Steam for the tool (like with Arx or Quake 2), normally it's able to find the working directory correctly. With this game, it tries to load the working directory as the steam directory (~/.steam/steam). I was able to work around it by placing a shell script in that directory that only changed directories to the actual path of the game and executed the script.

My packages.json entry looks like the following (with the workaround):

```
 "13230": {
        "game_name": "UT2004",
        "package": "dist.tar.xz",
        "command": "./ut2004route.sh"
}
```

Normally it'd look like this:
```
 "13230": {
        "game_name": "UT2004",
        "package": "dist.tar.xz",
        "command": "./run-ut2004.sh"
}
```

**Comment by Askmewho** [comment]
I keep having problems with make it work on linux, the texture files aren meet and they have different extensions than the requiered ones.

**Comment by Askmewho** [comment]
thjey are utx.uz2 and not .utx , obviusly if I rename them they fail

**Comment by rcgordon** [comment]
uz2 is compresssed (it's basically just zlib in a non-standard format). They should be uncompressed and then be .utx files in the Textures directory.

[I posted a simple C program to the ut2004 mailing list](https://icculus.org/pipermail/ut2004/2004-June/001012.html), back in the day, to decompress uz2 files.

**Comment by Askmewho** [comment]
Thanks, I ve already done that on windowd with a normal instalation now it runs great. Just no sound but I will need to do a workaround about why openal it's not working. Btw it was tested on rpi4 with box86 and great results !

**Comment by AsciiWolf** [comment]
Any news regarding the UT2004 support in Luxtorpeda? :-)

**Comment by dreamer** [comment]
@d10sfan packaged it - it's available in his fork: https://github.com/luxtorpeda-dev (I recommend testing his version of Luxtorpeda, but, I have not tested UT2004 myself - yet).

*d10sfan's fork is not an aggressive one - you can ask for support in official Luxtorpeda channel on Discord*
